<a href="https://colab.research.google.com/github/GRX782/cursoTaller1/blob/main/archivos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Curso Taller 1** - Dr-Ing. José Andrés Araya Obando

## **Práctica Archivos**

✅ Con esta práctica aprenderás a gestionar archivos de trabajo. <br> Esto lo vamos a estar utilizando muchas veces durante el desarrollo del curso.

Respecto a las celdas de texto ...

🥸 Cuando estamos escribiendo celdas de texto en Colab, en realidad usamos un lenguaje llamado **Markdown**, *el cual permite escribir texto con formato (negrita, listas, títulos, etc.) de manera sencilla.*

👉 El problema es que en Markdown, Enter no siempre significa un salto de línea (a veces sigue en el mismo párrafo). Para forzar que algo quede en la siguiente línea, podemos usar `<br>` = código de HTML que significa “break” (salto de línea).

## Gestión del archivo de trabajo

**Ejercicio 01 - Memoria**

**Paso 1.** Abre el CMD ⏩ (Windows + R, escribe cmd)

**Paso 2.** Pregúntale a tu computadora ¿cuál es la capacidad de memoria instalada?

In [ ]:
wmic computersystem get TotalPhysicalMemory

☹ ¿Recuerdas en qué unidades está representando esa respuesta?

**Paso 3.** Descarga el archivo "*memoria.txt*" de mi GitHub.

▶ Otra forma es que en el preview sencillamente copies el texto y lo pases a un archivo "*.txt*" nuevo (abre la aplicación "*Block de notas*")

**Paso 4.** abre el prompt de python desde la terminal

In [ ]:
python

**Paso 5.** Copia el texto en el prompt con "Ctrl+v" y le das "*Enter*"

⏰ Reflexiona: compara esta salida con la del paso 2. ¿Mejoró? ¿La respuesta te brinda más contexto?

**Paso 6.** Debes salir del prompt en la terminal CMD

In [ ]:
exit()

**Paso 7.** Ver la carpeta actual del CMD

In [ ]:
cd

**Paso 8.** Abre el archivo *"memoria.txt"* y dale salvar como "*memoria.py*"

✅ Consúltale a la IA ¿qué es un archivo ".py" y que relación tiene con los archivos ".txt"?

**Paso 9.** Guarda el archivo "memoria.py" en el directorio de trabajo del cmd que consultaste en el paso 7.

**Paso 10.** Abre el archivo ".py" desde el CMD. Para ello, ejecuta esta instrucción:

In [ ]:
python memoria.py

⚡Te acabo de enseñar algo muy útil. Cuando tenemos que correr una tarea en el CMD, para evitar estar escribiendo muchas lineas de código desde la terminal (lo cual resulta muy tedioso), puedes crear un ".py" a partir de un ".txt" y sencillamente invocarlo desde desde el CMD (no hace falta abrir el prompt)

✅ Los archivos que se ejecutan desde la terminal y contienen código para realizar tareas complejas se llaman "scripts". Pídele a la IA que te explique más acerca de esto (comprende qué son archivos ".bat", ".cmd",".ps1" y ".sh")

**Paso 11.** Vamos a consultar acerca de la memoria pero desde acá mismo, es decir, del Notebook

☣ Primero debes comprender que en Colab **estamos sobre Linux**, no sobre Windows (¡así como lo escuchaste!) <br> <br> ❌El comando wmic que aparece en la sentencia del paso 2 solo existe en Windows, así que el script memoria.py que preparaste con os.popen("wmic ...") no puede correr en Colab.   

In [43]:
# Intentar importar psutil; si no está, instalar en silencio y reintentar
try:
    import psutil, time
except ImportError:
    !pip install -q psutil
    import psutil, time

print("Consultando la memoria física instalada en el entorno de Colab...")

# Pequeño "reloj" animado
for c in "|/-\\":
    print(f"\rPor favor espere... {c}", end="")
    time.sleep(0.3)

# Obtener memoria total y disponible en GB
vm = psutil.virtual_memory()
total_gb = vm.total / (1024**3)
avail_gb = vm.available / (1024**3)
used_gb  = (vm.total - vm.available) / (1024**3)
pct_used = (vm.total - vm.available) / vm.total * 100

print("\nLa memoria total del entorno es: {:.2f} GB".format(total_gb))
print("Memoria en uso: {:.2f} GB ({:.2f}%)".format(used_gb, pct_used))
print("Memoria disponible: {:.2f} GB".format(avail_gb))


Consultando la memoria física instalada en el entorno de Colab...
Por favor espere... \
La memoria total del entorno es: 12.67 GB
Memoria en uso: 1.27 GB (10.05%)
Memoria disponible: 11.40 GB


❗El script que corrimos desde la terminal devuelve la memoria física total instalada en la computadora (la RAM completa que tienes en tu equipo). En cambio, el script en colab (psutil.virtual_memory) no está midiendo la RAM física real del equipo en donde corres, sino la "memoria virtual" asignada por el entorno de Google Colab. En otras palabras, Colab te da una máquina virtual con una cuota de memoria limitada y no tiene acceso al hardware real de tu notebook, solo a los recursos que asigna a la sesión.

⚠ En mi caso, pese a que tengo una RAM de 31.79 GB la RAM disponible dentro de la máquina virtual de Colab es de 12.67 GB. Acá entra tu análisis costo beneficio de trabajar directamente en un IDE o seguir por acá.

# **Archivos ".csv"**

**Paso 1.** Descarga de mi GitHub el archivo "*01-vivienda_2pisos_csv.csv*"

**Paso 2.** Guarda el archivo en el Google Drive con el nombre "01-vivienda_2pisos_csv_utf-8" y guárdalo como  CSV UTF-8 (delimitado por comas)(*.csv) en la lista de tipos de archivo.

**Paso 3.** Subir el archivo directamente

In [41]:
from google.colab import files
import io, pandas as pd

print("Selecciona tu archivo CSV…")
uploaded = files.upload()  # abre el selector; elige tu CSV

# Tomar el primer archivo subido
name = list(uploaded.keys())[0]

# Intento robusto de lectura con codificación
for enc in ("utf-8", "latin1", "cp1252"):
    try:
        df = pd.read_csv(io.BytesIO(uploaded[name]), encoding=enc)
        print(f"✅ Archivo leído con encoding: {enc}")
        break
    except UnicodeDecodeError:
        continue
else:
    raise UnicodeDecodeError("No se pudo leer el CSV con utf-8/latin1/cp1252. Verifica la codificación.")

print("Columnas del archivo:", list(df.columns))
display(df.head(20))


Selecciona tu archivo CSV…


Saving 01-vivienda_2pisos_csv_utf-8.csv to 01-vivienda_2pisos_csv_utf-8 (2).csv
✅ Archivo leído con encoding: utf-8
Columnas del archivo: ['Partida', 'Subpartida', 'Unidad', 'Cantidad', 'Observaciones/Supuestos']


Partida                                         Subpartida  \
0             Preliminares                 Limpieza y preparación del terreno   
1             Preliminares                               Trazado y nivelación   
2             Preliminares                 Servicios provisionales y andamios   
3              Cimentación         Excavación de zapatas y vigas de fundación   
4              Cimentación  Sustitución de suelo (10 cm) con material comp...   
5              Cimentación                       Concreto ciclópeo en zapatas   
6              Cimentación                     Vigas de fundación en concreto   
7              Cimentación                   Acero de refuerzo en cimentación   
8               Estructura                        Columnas de concreto armado   
9               Estructura                Entrepisos tipo vigueta y bovedilla   
10              Estructura                        Escalera de concreto armado   
11              Estructura                                        Viga corona   
12  Estructura - Aberturas           Mochetas/Jambas de puertas (0.15x0.20 m)   
13  Estructura - Aberturas                  Dinteles de puertas (0.15x0.20 m)   
14  Estructura - Aberturas          Mochetas/Jambas de ventanas (0.15x0.20 m)   
15  Estructura - Aberturas                 Dinteles de ventanas (0.15x0.20 m)   
16  Estructura - Aberturas           Acero de refuerzo en mochetas y dinteles   
17             Mampostería                     Bloque de concreto 15x20x40 cm   
18             Mampostería            Refuerzo horizontal y vertical en muros   
19             Mampostería                                    Mortero de pega   

   Unidad  Cantidad           Observaciones/Supuestos  
0      m²   150.000             Área de lote estimada  
1      m²   150.000   Según huella + áreas exteriores  
2     glb     1.000  Incluye seguridad y señalización  
3      m³    40.000             Profundidades típicas  
4      m³     7.500          0.10 m x 75 m² de huella  
5      m³    25.000                               NaN  
6      m³    12.000                               NaN  
7     ton     2.500                               NaN  
8      m³    20.000                               NaN  
9      m²    75.000          Incluye losa colaborante  
10     ud     1.000              12-14 huellas aprox.  
11     m³    25.000                               NaN  
12     m³     1.010       8 puertas; 2 jambas x 2.1 m  
13     m³     0.288            8 puertas; luz ? 1.2 m  
14     m³     1.008         12 ventanas; alto ? 1.4 m  
15     m³     0.540          12 ventanas; luz ? 1.5 m  
16    ton     0.300            Estimado 100-120 kg/m³  
17     m²   350.000                               NaN  
18    ton     0.800                               NaN  
19     m³    12.000                               NaN

**Paso 4.** Interactuar con los datos

In [42]:
# Ajusta estos nombres según tu CSV
COL_PARTIDA    = "Partida"
COL_SUBPARTIDA = "Subpartida"
COL_UNIDAD     = "Unidad"
COL_CANTIDAD   = "Cantidad"

def norm(x):
    return str(x).strip().lower()

p = input("Ingrese la PARTIDA: ").strip()
s = input("Ingrese la SUBPARTIDA: ").strip()

# Comparación tolerante a espacios y mayúsculas/minúsculas
mask = (df[COL_PARTIDA].map(norm) == norm(p)) & (df[COL_SUBPARTIDA].astype(str).map(norm) == norm(s))
res = df.loc[mask, [COL_PARTIDA, COL_SUBPARTIDA, COL_UNIDAD, COL_CANTIDAD]]

if res.empty:
    print("⚠️ No se encontró la combinación (partida, subpartida). Verifique valores y nombres de columna.")
else:
    for _, row in res.iterrows():
        print(f"✅ Unidad: {row[COL_UNIDAD]} | Cantidad: {row[COL_CANTIDAD]}")


Ingrese la PARTIDA: Estructura
Ingrese la SUBPARTIDA: Viga corona
✅ Unidad: m³ | Cantidad: 25.0
